src/main/java/org/opensearch/knn/index/memory/NativeMemoryCacheManager.java

```diff
@@ -0,0 +1,281 @@
+    }
+
+    /**
+     * Retrieves NativeMemoryAllocation associated with the nativeMemoryEntryContext.
+     *
+     * @param nativeMemoryEntryContext Context from which to get NativeMemoryAllocation
+     * @param isAbleToTriggerEviction Determines if getting this allocation can evict other entries
+     * @return NativeMemoryAllocation associated with nativeMemoryEntryContext
+     * @throws ExecutionException if there is an exception when loading from the cache
+     */
+    public NativeMemoryAllocation get(NativeMemoryEntryContext<?> nativeMemoryEntryContext,
+                                      boolean isAbleToTriggerEviction) throws ExecutionException {
+        if (!isAbleToTriggerEviction &&
+                !cache.asMap().containsKey(nativeMemoryEntryContext.getKey()) &&
+                maxWeight - getCacheWeightInKilobytes() - nativeMemoryEntryContext.calculateSizeInKb() <= 0
+        ) {
+            throw new NativeMemoryThrottleException("Failed to load \"" + nativeMemoryEntryContext.getKey() +
+                    "\" into memory.");
+        }
+
+        return cache.get(nativeMemoryEntryContext.getKey(), nativeMemoryEntryContext::load);
+    }
+
+    /**
+     * Invalidate entry from the cache.
+     *
+     * @param key Identifier of entry to invalidate
+     */
+    public void invalidate(String key) {
+        cache.invalidate(key);
+    }
+
+    /**
+     * Invalidate all entries in the cache.
+     */
+    public void invalidateAll() {
+        cache.invalidateAll();
+    }
+
+    /**
+     * Returns whether or not the capacity of the cache has been reached
+     *
+     * @return Boolean of whether cache limit has been reached
+     */
+    public Boolean isCacheCapacityReached() {
+        return cacheCapacityReached.get();
+    }
+
+    /**
+     * Sets cache capacity reached
+     *
+     * @param value Boolean value to set cache Capacity Reached to
+     */
+    public void setCacheCapacityReached(Boolean value) {
+        cacheCapacityReached.set(value);
+    }
+
+    /**
+     * Get the stats of all of the Elasticsearch indices currently loaded into the cache
+     *
+     * @return Map containing all of the Elasticsearch indices in the cache and their stats
+     */
+    public Map<String, Map<String, Object>> getIndicesCacheStats() {
+        Map<String, Map<String, Object>> statValues = new HashMap<>();
+        String indexName;
+        NativeMemoryAllocation.IndexAllocation indexAllocation;
+
+        for (Map.Entry<String, NativeMemoryAllocation> entry : cache.asMap().entrySet()) {
+
+            if (entry.getValue() instanceof NativeMemoryAllocation.IndexAllocation) {
+                indexAllocation = (NativeMemoryAllocation.IndexAllocation) entry.getValue();
+                indexName = indexAllocation.getOsIndexName();
+                statValues.putIfAbsent(indexName, new HashMap<>());
+
+                statValues.get(indexName).put(GRAPH_COUNT, ((Integer) statValues.get(indexName)
```

Comment:
This function is a little confusing. Each entry we are looping over corresponds to a knn segment file. A k-NN index will have multiple segment files. GRAPH_COUNT will correspond to how many graphs (same as segments) a k-NN index will have. If I did put if absent, it wouldn't be able to increment the value.

For the other 2, they use putIfAbsent because the function they call actually gets the value for the entire index.

I could add a graphCountForIndex function to make more consistent.

Reply:
What are the cons of adding a graphCountForIndex function?  Would it make it less efficient? I knows the pros is better readability.

Reply:
Right, it is slightly more efficient to increment, as opposed to having a separate function.

In this for loop, we are already looping over every segment in the cache. If we had the utility function, for each OpenSearch index that has a segment in the cache, we would need to loop over all of the segments in the cache and determine how many of those segments belong to that OpenSearch index.

That being said, I dont think it will be a noticeable performance difference, so I will side with your approach which has better readability.